examples/components.md

```diff
@@ -0,0 +1,727 @@
+    name: example
+  path: deployment.yaml
+EOF
+
+cat <<EOF >$EXT_DB/deployment.yaml
+- op: add
+  path: /spec/template/spec/volumes/0
+  value:
+    name: dbpass
+    secret:
+      secretName: dbpass
+- op: add
+  path: /spec/template/spec/containers/0/volumeMounts/0
+  value:
+    mountPath: /var/run/secrets/db/
+    name: dbpass
+EOF
+
+cat <<EOF >$EXT_DB/configmap.yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: conf
+data:
+  db.conf: |
+    endpoint=127.0.0.1:1234
+    name=app
+    user=admin
+    pass=/var/run/secrets/db/dbpass.txt
+EOF
+```
+
+Define an `ldap` component, that creates a `Secret` for the LDAP password
+and a new entry in the `ConfigMap`:
+
+```shell
+LDAP=$DEMO_HOME/components/ldap
+mkdir -p $LDAP
+
+cat <<EOF >$LDAP/kustomization.yaml
+kind: v1alpha1/KustomizationPatch
+
+secretGenerator:
+- name: ldappass
+  files:
+    - ldappass.txt
+
+patchesStrategicMerge:
+  - configmap.yaml
+
+patchesJson6902:
+- target:
+    group: apps
+    version: v1
+    kind: Deployment
+    name: example
+  path: deployment.yaml
+EOF
+
+cat <<EOF >$LDAP/deployment.yaml
+- op: add
+  path: /spec/template/spec/volumes/0
+  value:
+    name: ldappass
+    secret:
+      secretName: ldappass
+- op: add
+  path: /spec/template/spec/containers/0/volumeMounts/0
+  value:
+    mountPath: /var/run/secrets/ldap/
+    name: ldappass
+EOF
+
+cat <<EOF >$LDAP/configmap.yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: conf
+data:
+  ldap.conf: |
+    endpoint=ldap://ldap.example.com
+    bindDN=cn=admin,dc=example,dc=com
+    pass=/var/run/secrets/ldap/ldappass.txt
+EOF
+```
+
+Define a `recaptcha` component, that creates a `Secret` for the reCAPTCHA
+site/secret keys and a new entry in the `ConfigMap`:
+
+```shell
+RECAPTCHA=$DEMO_HOME/components/recaptcha
+mkdir -p $RECAPTCHA
+
+cat <<EOF >$RECAPTCHA/kustomization.yaml
+kind: v1alpha1/KustomizationPatch
+
+secretGenerator:
+- name: recaptcha
+  files:
+    - site_key.txt
+    - secret_key.txt
+
+patchesStrategicMerge:
+  - configmap.yaml
+
+patchesJson6902:
+- target:
+    group: apps
+    version: v1
+    kind: Deployment
+    name: example
+  path: deployment.yaml
+EOF
+
+cat <<EOF >$RECAPTCHA/deployment.yaml
+- op: add
+  path: /spec/template/spec/volumes/0
+  value:
+    name: recaptcha
+    secret:
+      secretName: recaptcha
+- op: add
+  path: /spec/template/spec/containers/0/volumeMounts/0
+  value:
+    mountPath: /var/run/secrets/recaptcha/
+    name: recaptcha
+EOF
+
+cat <<EOF >$RECAPTCHA/configmap.yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: conf
+data:
+  recaptcha.conf: |
+    enabled=true
+    site_key=/var/run/secrets/recaptcha/site_key.txt
+    secret_key=/var/run/secrets/recaptcha/secret_key.txt
+EOF
+```
+
+Define a `community` overlay, that bundles the external DB and reCAPTCHA
+components:
+
+```shell
+COMMUNITY=$DEMO_HOME/overlays/community
+mkdir -p $COMMUNITY
+
+cat <<EOF >$COMMUNITY/kustomization.yaml
+kind: Kustomization
+resources:
```

Comment:
Probably makes sense to have components separate from resources.

```yaml
components:
- ../../components/external_db
- ../../components/recaptcha
```

Reply:
Our approach was to stay as close as possible to the existing interface of kustomization files.

Since the `bases` keyword was also absorbed by `resources` quite recently we believe it is more straight forward to have the `resources` keyword cover all cases, including components. 
Aside, such a change might have implications in the size and complexity of #2168.

In any case, we are definitely open to iterate on this and reach a common understanding.

Reply:
A `Component` (aka `KustomizationPatch`) effectively become the parent resource of the resources that come before it in the `resources` list.  They could be moved to a separate `components` list, which I think could then be evaluated as if it was appended to the end of the `resources` list.  I guess this would make things a little clearer (and maybe opportunities for checking that only the correct types are in each list), and probably wouldn't make the code that much more complicated.  The disadvantage is that you wouldn't be able to add any resources to be evaluated after the components, but I can't think of any examples that would actually need that (and the examples don't), so maybe that isn't a real concern?

Reply:
I implemented support for a `components` list on a [separate branch in my fork](https://github.com/pgpx/kustomize/tree/feature-componentlist) - doesn't seem to bad, with most of the added complexity needed to limit each kind to their own list.  The code definitely could do with a slightly more extensive refactor if `components` is preferred, but I didn't do that now just to minimise the extent of the changes.

Reply:
I think this is reasonable for now.  It has an alpha apiVersion so we aren't commitment to keeping the API in this form.

Reply:
Should I merge that in to the main PR branch?  I still need to make the change so make it use alpha instead of beta.

Reply:
I rebased the [components branch](https://github.com/pgpx/kustomize/tree/feature-componentlist) and added `v1alpha1` support.